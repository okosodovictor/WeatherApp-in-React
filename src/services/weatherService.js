function weatherService() {
  const store = {
    Montreal: "https://www.meteomedia.com/api/obsdata/caqc0363/c?ref=rt",
    Toronto: "https://www.theweathernetwork.com/api/obsdata/caon0696/c?ref=rt"
  };

  return {
    getWeather: async function(city) {
      let url = store[city];
      let response = await fetch(url);
      return await response;
    }
  };
}

export default weatherService;
