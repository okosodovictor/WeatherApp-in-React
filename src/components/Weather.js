import React from "react";
import weatherService from "../services/weatherService";

class Weather extends React.Component {
  constructor(props) {
    super(props);
    this.state = { weather: {}, city: "" };
  }

  render() {
    const {
      updatetime,
      wxcondition,
      feels_like,
      temperature,
      temperature_unit,
      icon
    } = this.state.weather;
    return (
      <div className="container">
        <h2>City Weather</h2>
        <table className="table table-striped table-hover table-bordered">
          <tbody>
            <tr>
              <td>
                <strong>City Name</strong>
              </td>
              <td>{this.state.city}</td>
            </tr>
            <tr>
              <td>
                <strong> Condition</strong>
              </td>
              <td>
                {wxcondition}
                {icon && (
                  <img
                    src={
                      "https://s1.twnmm.com/images/en_ca/icons/wxicons_medium/" +
                      icon +
                      ".png"
                    }
                    alt=""
                  />
                )}
              </td>
            </tr>
            <tr>
              <td>
                <strong>Temperature</strong>
              </td>
              <td>
                {temperature}
                <span />{" "}
                {temperature_unit === "C" ? (
                  <span>&#x2103;</span>
                ) : (
                  <span>&#x2109;</span>
                )}
              </td>
            </tr>
            <tr>
              <td>
                <strong>Feels like</strong>
              </td>
              <td>
                {feels_like}{" "}
                {temperature_unit === "C" ? (
                  <span>&#x2103;</span>
                ) : (
                  <span>&#x2109;</span>
                )}
              </td>
            </tr>
            <tr>
              <td>
                <strong>Temperature Unit</strong>
              </td>
              <td>{temperature_unit}</td>
            </tr>
            <tr>
              <td>
                <strong>Date last updated time:</strong>
              </td>
              <td>{updatetime}</td>
            </tr>
          </tbody>
        </table>
      </div>
    );
  }

  async loadWeatherData(city) {
    try {
      let response = await weatherService().getWeather(city);
      return response.json();
    } catch (error) {
      console.log(error);
    }
  }

  async componentDidMount() {
    let city = this.props.match.params.city;
    let data = await this.loadWeatherData(city);
    this.setState({ weather: data, city: city });
  }

  async componentWillReceiveProps(nextProps) {
    let nextCity = nextProps.match.params.city;
    if (nextCity !== this.props.match.params.city) {
      let newData = await this.loadWeatherData(nextCity);
      this.setState({ weather: newData, city: nextCity });
    }
  }
}

export default Weather;
