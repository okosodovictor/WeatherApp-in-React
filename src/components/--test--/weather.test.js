import React from "react";
import Weather from "../Weather";
import { shallow } from "../../enzyme";
import weatherService from "../../services/weatherService";
describe("Weather tests", () => {
  beforeEach(function() {
    let expectedResponse = {
      lbl_updatetime: "Updated on",
      updatetime: "Sat Jul 13 7:15 AM",
      updatetime_stamp_gmt: "1563016500000",
      wxcondition: "Partly cloudy",
      icon: "3",
      inic:
        "iVBORw0KGgoAAAANSUhEUgAAAAcAAAAHCAMAAADzjKfhAAAAS1BMVEUAAADfqE/i1KyqqqruuFnkrlHCw8bssEjq2on74ZXux3P40XvMzM7Bp330wljh4eL3yGS9qo3p5uLusEbg4eDz8u//7Ib333n62zc3bsJYAAAAFHRSTlMAB/hNPS8UDP7tysWsi4iBdHBlGexLoNwAAAA2SURBVAjXFcrHEcAgDADBE9HZBBH6r5Rh3wtIZHPpcwL17sOUgD/n0dpPvIyqPuDza23YX4AFKzUBiDlmoFoAAAAASUVORK5CYII=",
      temperature: "19",
      feels_like: "19",
      temperature_unit: "C",
      placecode: "CAON0696"
    };

    global.fetch = jest
      .fn()
      .mockImplementation(() => Promise.resolve(expectedResponse));
  });

  it("renders without crashing", () => {
    let weatherPage = shallow(<Weather />);
  });

  it("Should render weather Image", () => {
    let weatherPage = shallow(<Weather />);
    const image = weatherPage.find("img");
    expect(image).toBeDefined();
  });

  it("calls componentDidMount", () => {
    jest.spyOn(Weather.prototype, "componentDidMount");
    const wrapper = shallow(<Weather />);
    expect(Weather.prototype.componentDidMount.mock.calls.length).toBe(1);
    expect(wrapper.find("table").length).toBe(1);
  });

  // Montreal:
  it("load Weather Data for Montreal", async function() {
    const response = await weatherService().getWeather("Montreal");
    expect(response).toMatchSnapshot();
  });

  // Toronto:
  it("load Weather Data for Toronto", async function() {
    const response = await weatherService().getWeather("Toronto");
    expect(response);
  });
});
