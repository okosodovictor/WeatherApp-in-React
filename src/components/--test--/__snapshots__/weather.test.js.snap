// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather tests load Weather Data for Montreal 1`] = `
Object {
  "feels_like": "19",
  "icon": "3",
  "inic": "iVBORw0KGgoAAAANSUhEUgAAAAcAAAAHCAMAAADzjKfhAAAAS1BMVEUAAADfqE/i1KyqqqruuFnkrlHCw8bssEjq2on74ZXux3P40XvMzM7Bp330wljh4eL3yGS9qo3p5uLusEbg4eDz8u//7Ib333n62zc3bsJYAAAAFHRSTlMAB/hNPS8UDP7tysWsi4iBdHBlGexLoNwAAAA2SURBVAjXFcrHEcAgDADBE9HZBBH6r5Rh3wtIZHPpcwL17sOUgD/n0dpPvIyqPuDza23YX4AFKzUBiDlmoFoAAAAASUVORK5CYII=",
  "lbl_updatetime": "Updated on",
  "placecode": "CAON0696",
  "temperature": "19",
  "temperature_unit": "C",
  "updatetime": "Sat Jul 13 7:15 AM",
  "updatetime_stamp_gmt": "1563016500000",
  "wxcondition": "Partly cloudy",
}
`;

exports[`Weather tests load Weather Data for Toronto 1`] = `
Response {
  "_bodyBlob": Blob {},
  "_bodyInit": Blob {},
  "headers": Headers {
    "map": Object {
      "cache-control": "max-age=27, s-maxage=300",
      "content-type": "text/html; charset=utf-8",
      "expires": "Sun, 14 Jul 2019 19:14:59 GMT",
    },
  },
  "ok": true,
  "status": 200,
  "statusText": "OK",
  "type": "default",
  "url": "https://www.meteomedia.com/api/obsdata/caqc0363/c?ref=rt",
}
`;
